client/utils/guidance.py

```diff
@@ -129,49 +139,131 @@ def show_guidance_msg(self, unique_id, input_lines, access_token, hash_key, guid
+                for misu in lst_mis_u:
+                    log.info("Updating the count of misu: %s by %s", misu, increment)
+                    countData[misu] = countData.get(misu, 0) + increment
+
+                for misu in lst_mis_u:
+                    log.debug("Misu: %s has count %s", misu, countData[misu])
+                    if countData[misu] >= wa_count_threshold:
+                        msg_info = lambda_info_misu(wa_details, misu)
+                        if msg_info:
+                            msg_id_set.add(msg_info)
+            else:
+                # Lookup the lst_mis_u of each wrong answer in the list of wrong
+                # answers related to the current wrong answer (lst_assess_num),
+                # using dictAssessNum2AssessId
+                assess_num_to_aid = self.guidance_json['dictAssessNum2AssessId']
+
+                # misu -> list of wrong answers for that
+                related_misu_tags_dict = {}
+
+                for related_num, related_resp in lst_assess_num:
+                    related_aid = assess_num_to_aid[related_num]
+                    # Get the lst_misu for this asssigmment
+                    related_info = self.guidance_json['dictAssessId2Info'].get(related_aid)
+                    if not related_info:
+                        log.info("Could not find related id: %s in info dict",
+                                 related_aid)
+                        continue
+                    related_wa_info = related_info['dictWA2DictInfo'].get(related_resp)
+
+                    if not related_info:
+                        log.info("Could not find response %s in %s info dict",
+                                 related_resp, related_aid)
+                        continue
+
+                    related_misu_list = related_wa_info['lstMisU']
+                    for misu in related_misu_list:
+                        existing_resps = related_misu_tags_dict.get(misu, [])
+                        # Add dictWA2DictInfo to list of responses for this misunderstanding.
+                        related_misu_tags_dict[misu] = existing_resps + [related_wa_info]
+                        # Increment countDict for each tag in the set of tags for each related resp
+                        countData[misu] = countData.get(misu, 0) + 1
+
+                    for misu, lst_wa_info in related_misu_tags_dict.items():
+                        if countData[misu] > wa_count_threshold:
+                            for wa_info in lst_wa_info:
+                                print(wa_info)
+                                msg_id_set.add(lambda_info_misu(wa_info, misu))
+
+        self.save_misUdata(answerDict, countData)
 
         if len(msg_id_set) == 0:
-            log.info(
-                "No messages to display. Most likely hasn't hit the wrong answer threshold")
+            log.info("No messages to display. Most likely hasn't hit the wrong answer threshold")
```

Comment:
Nitpick: Please remove the "Most likely..." line this is not true if the student is in the control group.